### PR TITLE
Slfast2

### PR DESCRIPTION
**Pull request by Akababa**
Faster SL worker for parsing large PGN files. It simply iterates through and adds one time every game in every pgn in play_data folder, instead of choosing one at random like in online learning.

### CODE REVIEW

**Comment by benediamond** [comment]
Hi @Akababa, this looks like good work. If you don't mind, may I incorporate these changes into own fork, where I have been developing some things? If they prove successful, I will create a separate pull request.

**Comment by Akababa** [comment]
Of course, I'm glad to hear that :) But wait, I actually found a serious bug involving a mismatch of formats between optimize and save_play_data.
Since my fork is significantly different from this repo now I need to test it some more. If you can help write some tests that would help me a lot!

**Comment by benediamond** [comment]
@Akababa, here is another question. I see that you are using multi-threading (at least to flush the buffer). I am wondering if you know: in `player_chess.py`, is using `asyncio.gather` and `loop.run_until_complete` enough to induce multi-threading? Or does it just swap between processes on a single CPU thread?

If the latter, then it seems we could get big speedups in MCTS by properly introducing multi-threading. What do you think?

**Comment by benediamond** [comment]
If you don't mind me asking, could you describe the bug?

**Comment by benediamond** [comment]
I'll add that in my fork, I am using a custom procedure to store play data and read it to training data (see [`convert_to_training_data`](https://github.com/benediamond/chess-alpha-zero/blob/aa9d41896be49f341271e892e5e33e6ac07be9be/src/chess_zero/worker/optimize.py#L153) in `optimize.py` and [`gather_features`](https://github.com/benediamond/chess-alpha-zero/blob/aa9d41896be49f341271e892e5e33e6ac07be9be/src/chess_zero/env/chess_env.py#L177) in `chess_env.py`). So I'm not sure if the problem will be comparable. But I will test it.

**Comment by Akababa** [comment]
Of course :) I'm not sure if it exists in this repo, but here's what happened in mine:

In save_game_data (2 places), a game buffer is saved as all of black's moves followed by all of white's moves. However, in convert_to_training data, it assumes the game moves are in order (white, black, white black)
There's some other stuff at https://github.com/Akababa/chess-alpha-zero/issues/1, and maybe a few more that I forgot about (I'll try to write them as I remember)

**Comment by benediamond** [comment]
Yes, I know the issue. You can fix it by replacing the beginning of `save_play_data` with:
```python
        data = [move for pair in zip(self.white.moves, self.black.moves) for move in pair]  # interleave the two lists
        if len(self.white.moves) > len(self.black.moves):
            data += [self.white.moves[-1]]  # tack on final move if white moved last
        self.buffer += data
```

**Comment by Akababa** [comment]
yeah I fixed that and put a few more asserts in for piece of mind.

I don't know if it properly uses multithreading but I would guess not since a similar concept of "event loops" exists in javascript which is single threaded.

**Comment by benediamond** [comment]
I see. that seems like a big problem, as far as speed.

**Comment by Akababa** [comment]
Yeah it is. I simplified and removed some stuff around in the original MCTS, in addition to caching legal_moves, but of course proper multi-threading beats asyncio.

**Comment by benediamond** [comment]
@Akababa, @Zeta36, unrelated question. Have you tried messing around with the constants? What value for `c_puct` do you think will lead to more exploration and fewer draws by repetition?

See, if`c_puct` is higher, then the penalty assigned to a visited edge (s,a) resulting from incrementing the denominator of U(s,a) (i.e., 1 +  N(s,a)) will be more significant relative to the action values Q(s,a).

Yet a higher `c_puct` will also lead to very high prior P(s,a)-values having a stronger effect, and overpowering the Q(s,a)-values.

**Comment by Akababa** [comment]
I haven't been able to train much yet (I was too impatient and learning rate too low), but you could probably just "cheat" by adjudicating the games based on material.
In the AZ paper they also use 2 extra planes per board position (so 14x8x8) to count repetitions so we could try that to make it more obvious to the model.

**Comment by benediamond** [comment]
Note that in the AlphaZero paper, they begin with a learning rate of 0.2, before dropping it to 0.02, 0.002, and 0.0002. (This is ten times higher than AlphaGo Zero.)

In my fork, I have implemented the 14x8x8 (actually 119x8x8) input plane feature. I'm trying supervised learning now; I hope to demonstrate its success as a proof of concept and then perhaps push it into this repository as a branch.

Samuel's original commit used a material-based adjudication like the one you suggest. But I have been trying to avoid using human heuristics...

**Comment by Akababa** [comment]
I wish you the best of luck :) In my opinion, it's either human heuristics, or superhuman computing power...

**Comment by benediamond** [comment]
Great, I look forward to seeing it.
